projects/pgai/pgai/_install/install.py

```diff
@@ -0,0 +1,152 @@
+    with files("pgai.data").joinpath("ai.sql").open(mode="r") as f:
+        sql = f.read()
+    sql = sql.replace("@extschema:vector@", vector_extension_schema)
+    sql = sql.replace("__version__", __version__)
+    return sql
+
+
+def _get_guc_vectorizer_url_sql() -> sql_lib.SQL:
+    return sql_lib.SQL("select pg_catalog.current_setting(%s, true) as val")
+
+
+def _get_vector_extension_schema_sql() -> sql_lib.SQL:
+    return sql_lib.SQL("""
+        select n.nspname
+        from pg_extension e
+        join pg_namespace n on n.oid = e.extnamespace
+        where e.extname = 'vector'
+    """)
+
+
+def verify_error_library_already_installed(
+    error_from_result: psycopg.errors.DuplicateObject,
+) -> bool:
+    if error_from_result.diag.message_primary is None:
+        return False
+    return (
+        "the pgai library has already been installed/upgraded"
+        in error_from_result.diag.message_primary
+    )
+
+
+async def ainstall(
+    db_url: str, vector_extension_schema: str | None = None, if_not_exists: bool = True
+) -> None:
+    """Asynchronously install the pgai library into a PostgreSQL database.
+
+    Args:
+        db_url: Database connection URL
+        vector_extension_schema: Schema where the vector extension is installed if it
+            doesn't exist. If None, then the vector extension will be installed in the
+            default schema (default: None)
+        if_not_exists: If True, ignore if library is already installed. If False,
+            raise error (default: True)
+
+    Raises:
+        psycopg.errors.DuplicateObject: If library is already installed and
+            if_not_exists=False
+    """
+
+    async with (
+        await psycopg.AsyncConnection.connect(db_url, autocommit=True) as conn,
+        conn.cursor() as cur,
+        conn.transaction(),
+    ):
+        if vector_extension_schema is None:
+            await conn.execute("CREATE EXTENSION IF NOT EXISTS vector")
+        else:
+            await conn.execute(
+                sql_lib.SQL(
+                    "CREATE EXTENSION IF NOT EXISTS vector WITH SCHEMA {}"
+                ).format(sql_lib.Literal(vector_extension_schema))
+            )
+
+        await cur.execute(_get_vector_extension_schema_sql())
+        result = await cur.fetchone()
+        if result is None or result[0] is None:
+            raise Exception("vector extension not installed")
+
+        sql = _get_sql(result[0])
+
+        # if we need to send a ping to an external url then
+        # we need to install the ai extension
+        await cur.execute(_get_guc_vectorizer_url_sql(), (GUC_VECTORIZER_URL,))
+        result = await cur.fetchone()
+        if result is not None and result[0] is not None:
+            await conn.execute("CREATE EXTENSION IF NOT EXISTS ai cascade")
+
+        try:
+            await conn.execute(sql)  # type: ignore
+        except psycopg.errors.DuplicateObject as error_from_result:
```

Comment:
What ends up throwing the duplicate object exception? Everything in `head.sql` uses `if not exists`, right?

Reply:
I have `raise exception 'the pgai library has already been installed/upgraded' using errcode = '42710';` in head.sql and code 42710 is the duplicate object error.